### PR TITLE
Add one example to run batch inference distributed on Ray

### PR DESCRIPTION
**Pull request by c21**
This PR is to add an example to run vLLM batch inference in a multi-nodes environment.
Ray Data is used here to orchestrate the workflow:
* Read from multiple input files on cloud storage.
* Start vLLM engines on multiple nodes, run inference on all input in parallel.
* Write the result and prompt back to cloud storage.

Tested w/ 58k number of prompts and `meta-llama/Llama-2-7b-chat-hf`. Whole job took 5 minutes on 10 L4 GPU nodes.

### CODE REVIEW

**Comment by zhe-thoughts** [comment]
This is a small and clean example. lgtm

cc @simon-mo

**Review by Yard1 — COMMENTED** [review]
Let's verify this works with tensor parallelism.

File: examples/offline_inference_distributed.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+"""
+This example shows how to use Ray Data for running offline batch inference
+distributively on a multi-nodes cluster.
+
+Learn more of Ray Data in https://docs.ray.io/en/latest/data/data.html
```
[the remark this answers is not in the document]
Comment (c21) [reply examples/offline_inference_distributed.py]: Updated.

File: examples/offline_inference_distributed.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+# Initialize Ray cluster with required dependencies.
+ray.init(
+    runtime_env={
+        "pip": [
+            "accelerate>=0.16.0",
+            "transformers>=4.26.0",
+            "numpy<1.24",  # remove when mlflow updates beyond 2.2
+            "torch",
+        ]
```
[the remark this answers is not in the document]
Comment (c21) [reply examples/offline_inference_distributed.py]: Sorry, removed.

File: examples/offline_inference_distributed.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    # Specify the number of GPUs required per LLM instance.
+    num_gpus=1,
+    # Specify the batch size for inference.
+    batch_size=32,
+)
+
+# Write inference output data out as Parquet files to S3.
+# Multiple files would be written to the output destination,
+# and each task would write one or more files separately.
+ds.write_parquet("s3://<your-output-bucket>")
```
[the remark this answers is not in the document]
Comment (c21) [reply examples/offline_inference_distributed.py]: Yes, make sense. Updated.

**Comment by c21** [comment]
>Let's verify this works with tensor parallelism.

@Yard1 - confirmed tensor parallelism works when testing `num_gpus=0` with `tensor_parallel_size=2`. Underlying 2 child actors (logical GPU=1 per each) got scheduled, and 2 GPUs are actually used. I added comment in PR for tensor parallelism, do not set `num_gpus` for now.

**Review by simon-mo — APPROVED** [review]

**Comment by c21** [comment]
Thank you @simon-mo and @Yard1 for review!

**Pull request merged**